numcodecs/shuffle.pyx

```diff
@@ -0,0 +1,100 @@
+# cython: embedsignature=True
+# cython: profile=False
+# cython: linetrace=False
+# cython: binding=False
+# cython: language_level=3
+
+import numpy as np
+from .compat import ensure_contiguous_ndarray
+from .abc import Codec
+
+cimport cython
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef void _doShuffle(const unsigned char[::1] src, unsigned char[::1] des, Py_ssize_t element_size) nogil:
+    cdef Py_ssize_t count, i, j, offset, byte_index
+    count = len(src) // element_size
+    for i in range(count):
+        offset = i*element_size
+        for byte_index in range(element_size):
+            j = byte_index*count + i
+            des[j] = src[offset + byte_index]
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef void _doUnshuffle(const unsigned char[::1] src, unsigned char[::1] des, Py_ssize_t element_size) nogil:
+    cdef Py_ssize_t count, i, j, offset, byte_index
+    count = len(src) // element_size
+    for i in range(element_size):
+        offset = i*count
+        for byte_index in range(count):
+            j = byte_index*element_size + i
+            des[j] = src[offset+byte_index]
+
+
+def _shuffle(element_size, buf, out):
+    if element_size <= 1:
+        out.view(buf.dtype)[:len(buf)] = buf[:]
+        return out  # no shuffling needed
+
+    buf_size = buf.nbytes
+    if buf_size % element_size != 0:
+        raise ValueError("Shuffle buffer is not an integer multiple of elementsize")
+
+    _doShuffle(buf.view("uint8"), out.view("uint8"), element_size)
+
+    return out
+
+
+def _unshuffle(element_size, buf, out):
+    if element_size <= 1:
+        out.view(buf.dtype)[:len(buf)] = buf[:]
+        return out  # no shuffling needed
+
+    buf_size = buf.nbytes
+    if buf_size % element_size != 0:
+        raise ValueError("Shuffle buffer is not an integer multiple of elementsize")
+
+    _doUnshuffle(buf.view("uint8"), out.view("uint8"), element_size)
+
+    return out
```

Comment:
I wonder if we can move this into the `Shuffle` codec. Would simplify the code here a bit and avoid the overhead of these intermediate functions

Reply:
Yep good point. refactored accrodingly